src/JustSaying/Messaging/Channels/MessageReceiveBuffer.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Amazon.SQS;
+using Amazon.SQS.Model;
+using JustSaying.AwsTools.MessageHandling;
+using JustSaying.Messaging.MessageProcessingStrategies;
+using JustSaying.Messaging.Middleware;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal class MessageReceiveBuffer : IMessageReceiveBuffer
+    {
+        private readonly Channel<IQueueMessageContext> _channel;
+        private readonly int _bufferLength;
+        private readonly ISqsQueue _sqsQueue;
+        private readonly MiddlewareBase<GetMessagesContext, IList<Message>> _sqsMiddleware;
+        private readonly IMessageMonitor _monitor;
+        private readonly ILogger _logger;
+
+        private readonly List<string> _requestMessageAttributeNames = new List<string>();
+
+        public ChannelReader<IQueueMessageContext> Reader => _channel.Reader;
+
+        public MessageReceiveBuffer(
+            int bufferLength,
+            ISqsQueue sqsQueue,
+            MiddlewareBase<GetMessagesContext, IList<Message>> sqsMiddleware,
+            IMessageMonitor monitor,
+            ILoggerFactory logger,
+            IMessageBackoffStrategy messageBackoffStrategy = null)
+        {
+            _bufferLength = bufferLength;
+            _sqsQueue = sqsQueue;
+            _sqsMiddleware = sqsMiddleware;
+            _monitor = monitor;
+            _logger = logger.CreateLogger<IMessageReceiveBuffer>();
+            _channel = Channel.CreateBounded<IQueueMessageContext>(bufferLength);
+
+            if (messageBackoffStrategy != null)
+            {
+                _requestMessageAttributeNames.Add(MessageSystemAttributeName.ApproximateReceiveCount);
+            }
+        }
+
+        /// <summary>
+        /// Starts the receive buffer until it's cancelled by the stopping token.
+        /// </summary>
+        /// <param name="stoppingToken"></param>
+        /// <returns>A task that completes when the cancellation token is fired.</returns>
+        public async Task Run(CancellationToken stoppingToken)
+        {
+            await Task.Yield();
+
+            ChannelWriter<IQueueMessageContext> writer = _channel.Writer;
+            try
+            {
+                while (true)
```

Comment:
I raised this question in the last PR, but reversed. I think it's advised to use throw rather than gracefully exiting. I prefer it for consistency, otherwise it depends on the timing of when you cancel as to how it behaves.
Would appreciate your view.

Reply:
The tests are rather nicely consistent now in that we _always_ throw when cancellation is fired, and we assert that is the case.

Reply:
I'm just wary of the `while (true)` construct in general.